drivers/display/display_gc9x01x.c

```diff
@@ -0,0 +1,745 @@
+			 const struct display_buffer_descriptor *desc, const void *buf)
+{
+	const struct gc9x01x_config *config = dev->config;
+	struct gc9x01x_data *data = dev->data;
+
+	int retval;
+	const uint8_t *write_data_start = (const uint8_t *)buf;
+	struct spi_buf tx_buf;
+	struct spi_buf_set tx_bufs;
+	uint16_t write_cnt;
+	uint16_t nbr_of_writes;
+	uint16_t write_h;
+
+	__ASSERT(desc->width <= desc->pitch, "Pitch is smaller than width");
+	__ASSERT((desc->pitch * data->bytes_per_pixel * desc->height) <= desc->buf_size,
+		 "Input buffer to small");
+
+	LOG_DBG("Writing %dx%d (w,h) @ %dx%d (x,y)", desc->width, desc->height, x, y);
+	retval = gc9x01x_set_mem_area(dev, x, y, desc->width, desc->height);
+	if (retval < 0) {
+		return retval;
+	}
+
+	if (desc->pitch > desc->width) {
+		write_h = 1U;
+		nbr_of_writes = desc->height;
+	} else {
+		write_h = desc->height;
+		nbr_of_writes = 1U;
+	}
+
+	retval = gc9x01x_transmit(dev, GC9X01X_CMD_MEMWR, write_data_start,
+				  desc->width * data->bytes_per_pixel * write_h);
+	if (retval < 0) {
+		return retval;
+	}
+
+	tx_bufs.buffers = &tx_buf;
+	tx_bufs.count = 1U;
+
+	write_data_start += desc->pitch * data->bytes_per_pixel;
+	for (write_cnt = 1U; write_cnt < nbr_of_writes; ++write_cnt) {
+		tx_buf.buf = (void *)write_data_start;
+		tx_buf.len = desc->width * data->bytes_per_pixel * write_h;
+
+		retval = spi_write_dt(&config->spi, &tx_bufs);
+		if (retval < 0) {
+			return retval;
+		}
+
+		write_data_start += desc->pitch * data->bytes_per_pixel;
+	}
+
+	return 0;
+}
+
+static int gc9x01x_read(const struct device *dev, const uint16_t x, const uint16_t y,
+			const struct display_buffer_descriptor *desc, void *buf)
+{
+	LOG_ERR("Reading not supported");
+	return -ENOTSUP;
+}
+
+static void *gc9x01x_get_framebuffer(const struct device *dev)
+{
+	LOG_ERR("Direct framebuffer access not supported");
+	return NULL;
+}
+
+static int gc9x01x_set_contrast(const struct device *dev, const uint8_t contrast)
+{
+	LOG_ERR("Set contrast not supported");
+	return -ENOTSUP;
+}
+
+static void gc9x01x_get_capabilities(const struct device *dev,
+				     struct display_capabilities *capabilities)
+{
+	struct gc9x01x_data *data = dev->data;
+	const struct gc9x01x_config *config = dev->config;
+
+	memset(capabilities, 0, sizeof(struct display_capabilities));
+
+	capabilities->supported_pixel_formats = PIXEL_FORMAT_RGB_565 | PIXEL_FORMAT_RGB_888;
+	capabilities->current_pixel_format = data->pixel_format;
+
+	if (data->orientation == DISPLAY_ORIENTATION_NORMAL ||
+	    data->orientation == DISPLAY_ORIENTATION_ROTATED_180) {
+		capabilities->x_resolution = config->x_resolution;
+		capabilities->y_resolution = config->y_resolution;
+	} else {
+		capabilities->x_resolution = config->y_resolution;
+		capabilities->y_resolution = config->x_resolution;
+	}
+
+	capabilities->current_orientation = data->orientation;
+}
+
+#ifdef CONFIG_PM_DEVICE
+static int gc9x01x_pm_action(const struct device *dev, enum pm_device_action action)
```

Comment:
I'm wondering whether PWM, GPIO and API also should be resumed and suspended.

Reply:
I am not an PM expert, but from what I understand in terms of device power management, every device needs to keep track of its used count (`pm_device_runtime_get`, `pm_device_runtime_put`). For the display driver the common layer will take care of this and this driver only implements/provides the actual execution of the according power management mode. Am I wrong, or what specifically do you want me implement here?

Reply:
I'm not an expert in this too. I thought that as these peripherals, PWM, GPIO and SPI, are initialized and controlled in the driver, the driver should be responsible for their PM but I might be wrong. It's better to do some research or ask someone who is familiar with this.

Reply:
I am unsure, what you want me to do here. I tested behaviour with a small sample which worked without problems.

As the display driver is not part of a power domain it can be suspended and resumed using the [device power management API](https://docs.zephyrproject.org/latest/services/pm/device.html#) directly.  Where do you see an issue here?